lib/posix/pthread_key.c

```diff
@@ -0,0 +1,178 @@
+	*key = (void *)new_key;
+
+	return 0;
+}
+
+/**
+ * @brief Delete a key for thread-specific data
+ *
+ * See IEEE 1003.1
+ */
+int pthread_key_delete(pthread_key_t key)
+{
+	pthread_key_obj *key_obj = (pthread_key_obj *)key;
+	pthread_key_data *key_data;
+	sys_dnode_t *node_l, *next_node_l;
+
+	k_sem_take(&pthread_key_sem, K_FOREVER);
+
+	/* Delete thread-specific elements associated with the key */
+	SYS_DLIST_FOR_EACH_NODE_SAFE(&(key_obj->key_data_l),
+			node_l, next_node_l) {
+
+		/* Remove the object from the list key_data_l */
+		key_data = (pthread_key_data *)
+			sys_dlist_get(&(key_obj->key_data_l));
+
+		/* Deallocate the object's memory */
+		k_free((void *)key_data);
+	}
+
+	k_free((void *)key_obj);
+
+	k_sem_give(&pthread_key_sem);
+
+	return 0;
+}
+
+/**
+ * @brief Associate a thread-specific value with a key
+ *
+ * See IEEE 1003.1
+ */
+int pthread_setspecific(pthread_key_t key, const void *value)
+{
+	pthread_key_obj *key_obj = (pthread_key_obj *)key;
+	struct posix_thread *thread = (struct posix_thread *)pthread_self();
+	pthread_key_data *key_data;
+	pthread_thread_data *thread_spec_data;
+	sys_dnode_t *node_l, *next_node_l;
+
+	/* Traverse the list of keys set by the thread, looking for key.
+	 * If the key is already in the list, re-assign its value.
+	 * Else add the key to the thread's list.
+	 */
+	k_sem_take(&pthread_key_sem, K_FOREVER);
+
+	SYS_DLIST_FOR_EACH_NODE_SAFE(&(thread->key_list), node_l, next_node_l) {
+
+			thread_spec_data = (pthread_thread_data *)node_l;
+
+			if (thread_spec_data->key == key_obj) {
+
+				/* Key is already present so
+				 * associate thread specific data
+				 */
+				thread_spec_data->spec_data = (void *)value;
+				goto out;
+			}
+	}
+
+	if (node_l == NULL) {
+		key_data = k_malloc(sizeof(pthread_key_data));
+
+		if (key_data == NULL) {
+			return ENOMEM;
+
+		} else {
+			/* Associate thread specific data, initialize new key */
+			key_data->thread_data.key = key;
+			key_data->thread_data.spec_data = (void *)value;
+
+			/* Append new thread key data to thread's key list */
+			sys_dlist_append((&thread->key_list),
+				(sys_dnode_t *)(&key_data->thread_data));
+
+			/* Append new key data to the key object's list */
+			sys_dlist_append(&(key_obj->key_data_l),
+					(sys_dnode_t *)key_data);
+		}
+	}
+
+out:
+	k_sem_give(&pthread_key_sem);
+
+	return 0;
+}
+
+/**
+ * @brief Get the thread-specific value associated with the key
+ *
+ * See IEEE 1003.1
+ */
+void *pthread_getspecific(pthread_key_t key)
+{
+	pthread_key_obj *key_obj = (pthread_key_obj *)key;
+	struct posix_thread *thread = (struct posix_thread *)pthread_self();
+	pthread_thread_data *thread_spec_data;
+	void *value = NULL;
+	sys_dnode_t *node_l, *next_node_l;
+
+	k_sem_take(&pthread_key_sem, K_FOREVER);
+
+	node_l = sys_dlist_peek_head(&(thread->key_list));
+
+	/* Traverse the list of keys set by the thread, looking for key */
+
+	SYS_DLIST_FOR_EACH_NODE_SAFE(&(thread->key_list), node_l, next_node_l) {
```

Comment:
You don't need to sue the SAFE variant of this macro if you're only traversing the list.  The SAFE variant is meant to be used when you're removing nodes as you traverse the list (as it saves the next pointer).